### PR TITLE
Support for multi-threaded sniffer

### PR DESCRIPTION
**Pull request by dgarske**
# Description

Support for multi-threaded sniffer.
Add support for atomic operations instead of mutex in wc_port.h.
ZD 14597
Author: @lealem47 

# Testing

 Tested with:

```
./configure --enable-sniffer && make check
./configure --enable-sniffer --enable-asynccrypt && make check
./configure --enable-sniffer --enable-asynccrypt CFLAGS=-"DWOLFSSL_SNIFFER_CHAIN_INPUT -DWOLFSSL_SNIFFER_WATCH -DTHREADED_SNIFFTEST" && make check
./configure --enable-sniffer CFLAGS=-DTHREADED_SNIFFTEST && make check
./configure --enable-sniffer --enable-asynccrypt CFLAGS=-DTHREADED_SNIFFTEST && make check
```

Also tested on QAT hardware.

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: src/sniffer.c [hunk 1]
```diff
@@ -626,8 +638,12 @@ void ssl_InitSniffer_ex(int devId)
 {
     wolfSSL_Init();
     wc_InitMutex(&ServerListMutex);
+#ifndef HAVE_C___ATOMIC
```
Comment (fencekicker) [inline src/sniffer.c:641]: Maybe you could put the #ifdef-ed locking and unlocking of SessionMutex under a LOCK_SESSION / UNLOCK_SESSION pair of macros?

File: wolfssl/wolfcrypt/wc_port.h [hunk 2]
```diff
@@ -278,37 +278,44 @@
...
-    while (0)
-
-#define wolfSSL_RefDec(ref, isZero, err)    \
-    do {                                    \
-        (ref)->count--;                     \
-        *(isZero) = ((ref)->count == 0);    \
-        *(err) = 0;                         \
-    }                                       \
-    while (0)
+#define wolfSSL_RefInc(ref, err)  ({        \
```
Comment (fencekicker) [inline wolfssl/wolfcrypt/wc_port.h:293]: This part (and wc_port.c) doesn't seem related to the sniffer changes? Or maybe I'm missing something? The code is newer than what we have.

Comment (dgarske) [reply wolfssl/wolfcrypt/wc_port.h:293]: Correct, not part of sniffer. I added detection for atomic and fixed this TODO. I will separate the commits.

**Comment by dgarske** [comment]
retest this please

**Comment by lealem47** [comment]
retest this please

File: src/sniffer.c (hunk 1 above)
Comment (lealem47) [reply src/sniffer.c:652]: Done

File: wolfssl/wolfcrypt/wc_port.h (hunk 2 above)
Comment (lealem47) [reply wolfssl/wolfcrypt/wc_port.h]: Done

**Review by dgarske — COMMENTED** [review]
The state of this looks good. @fencekicker , any feedback on this? I still plan to work with @lealem47 to add a threaded snifftest option.

**Comment by fencekicker** [comment]
As far as I can tell, this looks good. I'll still have to test this on our end, but it looks promising; the global ServerList was preventing us from running async code multi-threaded. The non-async mode already works multithreaded on our end.

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 3]
```diff
@@ -38,6 +38,10 @@
     #define WOLFSSL_SNIFFER
 #endif
 
+#ifdef THREADED_SNIFFTEST
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:41]: Can we just use `HAVE_PTHREAD` instead. Something like this:
```#if defined(HAVE_PTHREAD) && !defined(NO_THREADED_SNIFFTEST)
#define THREADED_SNIFFTEST
#endif
```

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 4]
```diff
@@ -569,6 +580,320 @@ static int SnifferAsyncPollQueue(byte** data, char* err, SSLInfo* sslInfo,
 }
 #endif /* WOLFSSL_ASYNC_CRYPT */
 
+#ifdef THREADED_SNIFFTEST
+
+/* Change to increase number of threads */
+const int workerThreadCount = 3;
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:586]: Please make this an overridable build macro:
```
#ifndef THREADED_SNIFFTEST_WORKERS
#define THREADED_SNIFFTEST_WORKERS 3
#endif

static const int workThreadCount = THREADED_SNIFFTEST_WORKERS;
```

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 5]
```diff
@@ -569,6 +580,320 @@ static int SnifferAsyncPollQueue(byte** data, char* err, SSLInfo* sslInfo,
...
+static int wm_SemUnlock(wm_Sem *s){
+    pthread_mutex_lock(&s->mutex);
+    s->lockCount--;
+    pthread_cond_signal(&s->cond);
+    pthread_mutex_unlock(&s->mutex);
+    return 0;
+}
+
+typedef struct SnifferWorker {
+    SnifferPacket *head; /* head for doubly-linked list of sniffer packtets */
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:623]: Spelling packtets

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 6]
```diff
@@ -569,6 +580,320 @@ static int SnifferAsyncPollQueue(byte** data, char* err, SSLInfo* sslInfo,
...
+{
+    SnifferPacket* newEntry;
+    byte*          packet;
+    int            length;
+
+#ifdef WOLFSSL_SNIFFER_CHAIN_INPUT
+    if (isChain) {
+        struct iovec* vchain = (struct iovec*)chain;
+        length = 0;
+        for (i = 0; i < chainSz; i++)
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:684]: Please test with `WOLFSSL_SNIFFER_CHAIN_INPUT`. `i` is not defined.

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 7]
```diff
@@ -569,6 +580,320 @@ static int SnifferAsyncPollQueue(byte** data, char* err, SSLInfo* sslInfo,
...
+        packet = (byte*)vchain[0].iov_base;
+    }
+    else
+#endif
+    {
+        packet = (byte*)chain;
+        length = chainSz;
+    }
+
+    newEntry = (SnifferPacket*)XMALLOC(sizeof(SnifferPacket), NULL,
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:695]: Consider single allocation `sizeof(SnifferPacket) + length`.

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 3 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c:41]: And make it the default case whenever pthread is available?

Comment (dgarske) [reply sslSniffer/sslSnifferTest/snifftest.c:41]: Yes if you add this above then it will do that:

```
#if defined(HAVE_PTHREAD) && !defined(NO_THREADED_SNIFFTEST)
#define THREADED_SNIFFTEST
#endif
```

Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c]: Done

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 4 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c]: Done

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 5 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c]: Fixed

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 6 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c]: Fixed and tested with `WOLFSSL_SNIFFER_CHAIN_INPUT`

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 7 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c:685]: This proved to be troublesome as I need to be able to free the ` SnifferPacket*` and `byte* packet` at different times.

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 8]
```diff
@@ -71,6 +78,8 @@ int main(void)
     #include <sys/socket.h>    /* AF_INET */
     #include <arpa/inet.h>
     #include <netinet/in.h>
+#else
+    #include <stdint.h>        /* uint8_t */
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:82]: Please use `byte` and remove include of stdint.h.

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 9]
```diff
@@ -469,17 +478,21 @@ static void show_usage(void)
...
-
 typedef struct SnifferPacket {
     byte* packet;
     int   length;
     int   lastRet;
     int   packetNumber;
+#ifdef THREADED_SNIFFTEST
+    struct SnifferPacket* next;
+    struct SnifferPacket* prev;
+    int    valid;
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:489]: What is `valid` used for? Seems like a workaround for something. Can you just check if `packet == NULL`? Consider better name if really required or making volatile.

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 10]
```diff
@@ -569,6 +583,323 @@ static int SnifferAsyncPollQueue(byte** data, char* err, SSLInfo* sslInfo,
...
+    /* decode packet */
+#ifdef WOLFSSL_ASYNC_CRYPT
+    /* For async call the original API again with same data,
+        * or call with different sessions for multiple concurrent
+        * stream processing */
+    ret = ssl_DecodePacketAsync(chain, chainSz, isChain, &data, err,
+        &sslInfo, NULL);
+
+    /* WC_PENDING_E: Hardware is processing or stream is blocked
+        *               (waiting on WC_PENDING_E) */
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:767]: Ident off

**Review by dgarske — COMMENTED** [review]
Fails with:

```
./configure --enable-sniffer && make

% ./scripts/sniffer-testsuite.test
sniffer features: key_callback tls_v13 tls_v12 static_ephemeral extended_master rsa dh ecc rsa_static dh_static
sniffer features: key_callback tls_v13 tls_v12 static_ephemeral extended_master rsa dh ecc rsa_static dh_static
sniffer features: key_callback tls_v13 tls_v12 static_ephemeral extended_master rsa dh ecc rsa_static dh_static
sniffer features: key_callback tls_v13 tls_v12 static_ephemeral extended_master rsa dh ecc rsa_static dh_static
sniffer features: key_callback tls_v13 tls_v12 static_ephemeral extended_master rsa dh ecc rsa_static dh_static
sniffer features: key_callback tls_v13 tls_v12 static_ephemeral extended_master rsa dh ecc rsa_static dh_static

Staring snifftest on testsuite.pcap...

snifftest 5.5.3
sniffer features: key_callback tls_v13 tls_v12 static_ephemeral extended_master rsa dh ecc rsa_static dh_static

SSL App Data(29:13):hello cyassl!
SSL App Data(31:23):I hear you fa shizzle!.
snifftest(65887,0x700001898000) malloc: *** error for object 0xffffffffa60fce08: pointer being freed was not allocated
snifftest(65887,0x700001898000) malloc: *** set a breakpoint in malloc_error_break to debug
./scripts/sniffer-testsuite.test: line 74: 65887 Abort trap: 6           ./sslSniffer/sslSnifferTest/snifftest ./scripts/sniffer-static-rsa.pcap ./certs/server-key.pem 127.0.0.1 11111

snifftest static RSA failed
```

If `--enable-asynccrypt` is required for sniffer threading then please gate enabling of the threading feature based on this.

File: sslSniffer/sslSnifferTest/snifftest.c [hunk 11]
```diff
@@ -569,6 +583,323 @@ static int SnifferAsyncPollQueue(byte** data, char* err, SSLInfo* sslInfo,
 }
 #endif /* WOLFSSL_ASYNC_CRYPT */
 
+#ifdef THREADED_SNIFFTEST
+
+#ifndef THREADED_SNIFFTEST_WORKERS
+    /* Change to increase number of threads */
+    #define THREADED_SNIFFTEST_WORKERS 4
```
Comment (dgarske) [inline sslSniffer/sslSnifferTest/snifftest.c:590]: Add way to supply this at run-time.

File: src/sniffer.c [hunk 12]
```diff
@@ -628,9 +645,13 @@ static void UpdateMissedDataSessions(void)
 void ssl_InitSniffer_ex(int devId)
 {
     wolfSSL_Init();
+#ifndef HAVE_C___ATOMIC
     wc_InitMutex(&ServerListMutex);
     wc_InitMutex(&SessionMutex);
+#endif
+#ifndef WOLFSSL_SNIFFER_NO_RECOVERY
     wc_InitMutex(&RecoveryMutex);
+#endif
```
Comment (dgarske) [inline src/sniffer.c:654]: Need to update the async open section to something like this:

```
#ifdef WOLFSSL_ASYNC_CRYPT
#ifndef WC_NO_ASYNC_THREADING
    ThreadData* threadData = (ThreadData*)args;

    if (wolfAsync_DevOpenThread(&devId, &threadData->thread_id) < 0)
#else
    if (wolfAsync_DevOpen(&devId) < 0)
#endif
    {
        printf("%sAsync device open failed\n%sRunning without async\n",
               err_prefix, err_prefix);
    }
#endif /* WOLFSSL_ASYNC_CRYPT */
```

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 10 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c]: Fixed

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 9 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c]: Yeah it's a workaround for the threads that might not receive packets. Renamed to `placeholder`.

File: sslSniffer/sslSnifferTest/snifftest.c (hunk 8 above)
Comment (lealem47) [reply sslSniffer/sslSnifferTest/snifftest.c]: Done

File: src/sniffer.c (hunk 12 above)
Comment (lealem47) [reply src/sniffer.c:654]: Added

**Review by dgarske — COMMENTED** [review]
Seeing some leaks with the threaded sniffer branch, even though it works fine. I can’t tell if these are real of it is is just the `wolfSSL_CTX_free(ctx)` not getting called.

```
./configure CC="clang -fsanitize=address,undefined -g" --enable-sniffer CFLAGS="-DDEBUG_SNIFFER" && make
./scripts/sniffer-testsuite.test
=================================================================
==2766653==ERROR: LeakSanitizer: detected memory leaks

Direct leak of 672 byte(s) in 4 object(s) allocated from:
    #0 0x49656d in malloc (/home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/.libs/snifftest+0x49656d)
    #1 0x7f5d54d1099b in SetNamedPrivateKey /home/davidgarske/GitHub/wolfssl/src/sniffer.c:1780:35
    #2 0x7f5d54d11509 in ssl_SetEphemeralKey /home/davidgarske/GitHub/wolfssl/src/sniffer.c:2015:11
    #3 0x4c9f2e in load_key /home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/snifftest.c:355:15
    #4 0x4ca2f2 in snifferWorker /home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/snifftest.c:849:5
    #5 0x7f5d548eb608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477:8

Indirect leak of 6596 byte(s) in 4 object(s) allocated from:
    #0 0x49656d in malloc (/home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/.libs/snifftest+0x49656d)
    #1 0x7f5d54b907f4 in AllocDer /home/davidgarske/GitHub/wolfssl/wolfcrypt/src/asn.c:22435:29
    #2 0x7f5d54ca0a36 in ProcessBuffer /home/davidgarske/GitHub/wolfssl/src/ssl.c:6476:15
    #3 0x7f5d54ca8416 in ProcessFile /home/davidgarske/GitHub/wolfssl/src/ssl.c:7973:19
    #4 0x7f5d54ca8f4e in wolfSSL_CTX_use_PrivateKey_file /home/davidgarske/GitHub/wolfssl/src/ssl.c:8728:9

Indirect leak of 6596 byte(s) in 4 object(s) allocated from:
    #0 0x49656d in malloc (/home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/.libs/snifftest+0x49656d)
    #1 0x7f5d54b907f4 in AllocDer /home/davidgarske/GitHub/wolfssl/wolfcrypt/src/asn.c:22435:29
    #2 0x7f5d54cb999e in SetStaticEphemeralKey /home/davidgarske/GitHub/wolfssl/src/ssl.c:38095:19
    #3 0x7f5d54cb96e6 in wolfSSL_CTX_set_ephemeral_key /home/davidgarske/GitHub/wolfssl/src/ssl.c:38249:12

Indirect leak of 1344 byte(s) in 4 object(s) allocated from:
    #0 0x49656d in malloc (/home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/.libs/snifftest+0x49656d)
    #1 0x7f5d54c90a1c in wolfSSL_CTX_new_ex /home/davidgarske/GitHub/wolfssl/src/ssl.c:514:25
    #2 0x7f5d54d11509 in ssl_SetEphemeralKey /home/davidgarske/GitHub/wolfssl/src/sniffer.c:2015:11
    #3 0x4c9f2e in load_key /home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/snifftest.c:355:15
    #4 0x4ca2f2 in snifferWorker /home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/snifftest.c:849:5
    #5 0x7f5d548eb608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477:8

Indirect leak of 1056 byte(s) in 4 object(s) allocated from:
    #0 0x49656d in malloc (/home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/.libs/snifftest+0x49656d)
    #1 0x7f5d54c9d044 in wolfSSL_CertManagerNew_ex /home/davidgarske/GitHub/wolfssl/src/ssl.c:4303:34

Indirect leak of 16 byte(s) in 4 object(s) allocated from:
    #0 0x49656d in malloc (/home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/.libs/snifftest+0x49656d)
    #1 0x7f5d54cae12d in wolfSSLv23_client_method_ex /home/davidgarske/GitHub/wolfssl/src/ssl.c:12531:49
    #2 0x7f5d54cae12d in wolfSSLv23_client_method /home/davidgarske/GitHub/wolfssl/src/ssl.c:12526:16
    #3 0x7f5d54d11509 in ssl_SetEphemeralKey /home/davidgarske/GitHub/wolfssl/src/sniffer.c:2015:11
    #4 0x4c9f2e in load_key /home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/snifftest.c:355:15
    #5 0x4ca2f2 in snifferWorker /home/davidgarske/GitHub/wolfssl/sslSniffer/sslSnifferTest/snifftest.c:849:5
    #6 0x7f5d548eb608 in start_thread /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:477:8

SUMMARY: AddressSanitizer: 16280 byte(s) leaked in 24 allocation(s).

snifftest static RSA failed
```

**Comment by dgarske** [comment]
Retest this please

**Review by dgarske — COMMENTED** [review]
Approved. Over to @SparkiDev for final review

File: wolfssl/wolfcrypt/wc_port.h [hunk 13]
```diff
@@ -284,37 +284,44 @@
...
 
 #ifdef SINGLE_THREADED
-#define wolfSSL_RefInit(ref, err)           \
-    do {                                    \
-        (ref)->count = 1;                   \
-        *(err) = 0;                         \
-    }                                       \
-    while (0)
-
+#define wolfSSL_RefInit(ref, err) do {      \
```
Comment (SparkiDev) [inline wolfssl/wolfcrypt/wc_port.h:294]: Follow the formatting as it was.

File: wolfssl/wolfcrypt/wc_port.h [hunk 14]
```diff
@@ -284,37 +284,44 @@
...
+    (ref)->count = 1;                       \
+    *(err) = 0;                             \
+} while(0);
```
Comment (SparkiDev) [inline wolfssl/wolfcrypt/wc_port.h:297]: No semi-colon on end of while statements.

**Review by SparkiDev — CHANGES_REQUESTED** [review]
Tested performance of atomics and they are faster.

**Review by SparkiDev — APPROVED** [review]
